### DISCUSSION TITLE
Fetching product attributes for filters on a category page.

### DISCUSSION THREAD

**Discussion by Kealtie** [swellstores/community#470] [Help] [answered]
Hello,

Our team has been struggling with an efficiënt way to fetch the necessary data to display an be able to apply filters for our product category page. Currently, we have to do multiple requests (based on the amount of products related to a category) to fetch all the different attribute values the products might contain. I'll add a screenshot of this page for a visual explanation.

<img width="757" alt="image" src="https://github.com/swellstores/community/assets/7847482/284373c1-39a3-46f6-aac3-141ca007d4c9">
In the image above you see two filterables on the left hand side (Lengte (mm)) and (Breedte (mm)) (meaning length and width in Dutch). The options being displayed under those are fetched throughout all these requests (in this case we have 891 products under this category) meaning we have to do 9 requests in total to fetch all the possible attribute values.

We do this with the following function
`export async function fetchSwellCollectionRecursive<ReturnType>(request: (index: number, perPage: number) => Promise<SwellCollection<ReturnType>>): Promise<Array<ReturnType>> {
    const perPage = 100; // 100 is the maximum per page

    const firstFetched = await request(1, perPage);
    let allResults = firstFetched.results;

    if (firstFetched.count > perPage) {
        const requests = [];

        for (let i = 1; (i - 1) * perPage <= firstFetched.count; i++) {
            requests.push(request(i, perPage));
        }

        const otherResponses = await Promise.all(requests);
        const otherResults = otherResponses.flatMap((result) => result.results);

        allResults = [
            ...allResults,
            ...otherResults,
        ];
    }

    return allResults;
}
`

My question is, how do you guys fetch those attributes and is there a more efficient way? Since for some category pages the requests currently take up to 12 seconds, which ofcourse is not production worthy.

Greetings,
Jordy

**Comment by ericingram** [comment]
Have you looked at the product filters method in swell-js? https://developers.swell.is/frontend-api/products#list-filtering

**Reply by Kealtie** [reply]
Yes, we have. But this only fetches the filters for the given products. Meaning I still have to do 9 requests to fetch all 891 products first before being able to fetch the filters, or am I missing some option here?

**Reply by logeshswell** [reply, marked as answer]
@Kealtie Yes, you're right. This is the only possible way at the moment if you don't have static filters.

**Reply by logeshswell** [reply]
Let me check with my team and see if we can improve the experience of similar use cases containing a larger number of products.

**Reply by Kealtie** [reply]
@logeshswell Alright, sounds great! When can we expect you to get back on this? This issue results in us not being able to go into production for our customer. We'd have to do alot of calls for our primary categories which contain a large amount of products (about 1k+).

**Reply by logeshswell** [reply]
I can confirm that it won't be an immediate fix, but I will keep you posted if I receive an update.

**Reply by Kealtie** [reply]
Thank you!
